HTML Lesson 1: Optional homework (#36)
Code-Your-Future/syllabus: docs/react/week-3/lesson.md

  }
```

8. Now `App` **does** render `MartianImageFetcher` and passes the `date` state as a prop named `photoDate`
9. The `MartianImageFetcher` function component is called for the first time
10. `useState` knows that we haven't got any `imgSrc` state so initialises it to `null`

> 11. We queue an effect, which will run after we render for the first time

Reviewer: On the next line, I think it's the first time we're explained that returning `null` renders nothing. Could we explain this earlier (maybe in the ternary/&& section), so that this isn't _another_ piece of new information in this already quite complex set of bullet points?

Reviewer 2: Thanks, good catch. Originally it was mentioned in the conditional rendering section, but I guess I removed it when I was streamlining that section. I'll see if I can find a good place for it.

Reviewer 2: I ended up adding a note to the "The timeline of this component" bit.

        ...
        <button onClick={handle2020Click}>Fetch image for 2020</button>
        ...

>         {date ? <MartianImageFetcher photoDate={date} /> : null}

Reviewer: I'm worried that, having not explained any of this (i.e. the dependencies array, or even what a dependency is), bullet point 18 is confusing, and may leave students feeling like this is more complicated than it is just because they're not familiar with the terminology - particularly as the phrasing of the bullet point suggests to me that the reader _should_ already know this, but we didn't explain any of these concepts to them yet.

Perhaps we could add something to the end of the previous "The timeline of this component" para, like "When the response is received, update the state and re-render so that we can show the data on-screen. The way we've called it, `useEffect` _only_ calls its callback function the first time a component renders - when we do the re-render to show the data on-screen, we will _not_ do another `fetch`."

Then we could rephrase this bullet point as:
This time, we do **not** queue an effect. React doesn't know that it should - remember, the point of `useEffect` is to not call the callback function every time the component renders, so we only queued the effect on the **first** render!

Reviewer 2: Thanks, this is good feedback.

I'm a bit in two minds here: I was intentionally avoiding the deps array in the "The timeline of this component" para, because I wanted to get across the "loop" concept of setting state inside the fetch first. I didn't want to get into too much detail that distracts from this point. The "walkthrough" section was designed to introduce the deps array separately.

However, I do agree that it makes point 18 too confusing and out-of-context. I therefore added a sentence below the "The timeline of this component" para which calls out that the effect wasn't run a second time, but says that it'll be covered later.

I also reworded point 18 a bit so that it doesn't explicitly call it the "dependencies array". Feedback welcome

Reviewer: I think this looks great, a huge improvement!

One thing we may want to do is also add a `console.log("Fetching data from NASA");` line in the `useEffect` callback before the `fetch`, so that we can more obviously demonstrate that the callback isn't being called again. As things currently stand, we say "You might notice that even though we re-rendered, we did not run the useEffect a second time", but we haven't really _shown_ the students that the callback isn't called again (for all they know it is called again, and is returning the same result so not re-rendering, or even _is_ re-rendering just very quickly, or something) - being able to say "The long only happened once, so the function was only called once" could be a nice demonstration.

Another brief thought: I _think_ we've ended up in a place where we don't actually define the word "dependency" any more... We have the text:
```
Because of this behaviour, the second argument to useEffect is called the dependencies argument. We use it whenever we have something in our effect function that depends on a variable outside of the effect function.
```
But I think the first of those sentences pre-supposes being familiar with the concept of a dependency... Maybe we could add a first sentence to the "`useEffect` dependencies array" section, something like: "Our callback function uses the value of the `photoDate` `prop` - what the function does will depend on what value of `photoData` was used, so we call `photoData` a _dependency_ of the calllback function."?

Reviewer 2: EDIT: I added the log, that's a good idea.

> I think we've ended up in a place where we don't actually define the word "dependency" any more...

We have the hidden answer to Exercise D, which has this text:

"We call this a _dependency_ of the effect. Any variables which are used inside the `useEffect` callback function are dependencies."

I'm unsure if this is obvious enough that everyone would find it though?

I originally hid it because it would be trivial to answer Exercise D by just scrolling down a bit. I could just un-hide it and assume that we could have discussion anyway?

  }
```

| Exercise D (estimate: 5 min)                                                                          |
| :---------------------------------------------------------------------------------------------------- |
| 1. Did you spot where the bug was? Discuss with a group of 2 - 3 students where you think the bug is. |
| 2. Report back to the rest of the class where you think the bug happened.                             |

<details>
<summary>Click here to reveal the answer!</summary>

> The key that the `useEffect` in `MartianImageFetcher` is **only run once**. This is because we told React that the queue should be queued on the first render only. However, as we saw, sometimes you need the effect to run again when some data changes. In this case the `date` prop, changed from `"2019-01-01"` to `"2020-01-01"`, meaning that we have to fetch data different data. We call this a _dependency_ of the effect.

Reviewer: ```suggestion
The key that the `useEffect` in `MartianImageFetcher` is **only run once**. This is because we told React that the queue should be queued on the first render only. However, as we saw, sometimes you need the effect to run again when some data changes. In this case the `date` prop, changed from `"2019-01-01"` to `"2020-01-01"`, meaning that we have to fetch data different data. We call this a _dependency_ of the effect - any of our variables which we use in the callback function we gave to `useEffect` is a dependency.
```

Reviewer 2: I reworded this suggestion a bit, but thanks!

  useEffect(() => {
    ...
  }, [props.photoDate]);

  ...
}
```

> Now when the `photoDate` prop changes, React knows that the effect must be run again, this time with the 2020 data. Because of this behaviour, the second argument to `useEffect` is called the _dependencies argument_. We use it whenever we have something in our effect that _depends_ on a variable outside of the effect function.

Reviewer: ```suggestion
Now when the `photoDate` prop changes, React knows that the effect must be run again, this time with the 2020 date. Because of this behaviour, the second argument to `useEffect` is called the _dependencies argument_. We use it whenever we have something in our effect function that _depends_ on a variable outside of the effect function.
```

| 5. Add an `<input type="text" />` element above the `<p>` element.                                                                                                                                      |
| 6. Set the `value` attribute of the `<input>` to the `city` state.                                                                                                                                      |
| 7. Create a function within the component called `updateCity`. Pass this function to the `onChange` event handler.                                                                                      |
| 8. Change the `updateCity` component so that it has a parameter named `event`.                                                                                                                          |
| 9. Add a `console.log` to inspect the value of `event.target.value`. What happens when you type in the input?                                                                                           |

> | 10. Using `setCity`, update the city state to what what is typed in the input box.                                                                                                                      |

Reviewer: ```suggestion
| 10. Using `setCity`, update the city state to what was typed in the input box.                                                                                                                      |
```

}
```

> We now have three different inputs named `username`, `email` and `password`, and we keep each entered value in state as a field in an object.

Reviewer: I don't see an object? It looks like we're using three state hooks in a function? Maybe reword?

> The `handleChange` function is reused to keep track of changes for **all** the values. To be able to tell which `<input>` is being updated, we check the `event.target.name` field. This corresponds to the `name` attribute on the `<input>` (e.g. `<input name="username">`).

Reviewer: ```suggestion
The `handleChange` function is reused to keep track of changes for **all** the form fields. To be able to tell which `<input>` is being updated, we check the `event.target.name` field. This corresponds to the `name` attribute on the `<input>` (e.g. `<input name="username">`).
```

Based on this value, we then decide which state to update:

```js

> if (event.target.name === "username") {

Reviewer: Maybe this lesson is already long enough, but this could be a really nice opportunity to show a contrasting implementation where we use arrow functions as event handlers, and to remind the students once again about needing function references. Something like:

```jsx
<input onClick={event => setUsername(event.target.value)} />
<input onClick={event => setEmail(event.target.value)} />
<input onClick={event => setPassword(event.target.value)} />
```

This is something I've seen them try to do, and get stuck on a lot. Very frequently they try to write;
```jsx
<button onClick={incrementCounter()} />
```
and get confused when it doesn't work.

The fact that the single event handler function is a bit involved and fiddly, and these little arrow functions are concise, leads to a nice contrasting (and then we could maybe say something like "But if you're doing some validation or something, maybe the single function is a better approach because you can centralise logic"). Contrasting approaches, and thinking about why one is better than the other, isn't something we do much of, and would be great to do more of :)

Reviewer 2: > Very frequently they try to write;

That is why we have such an extensive event handling section in week 2 🙂 (IME most React tutorials hardly touch on this).

But I do agree that it's one of the most common problems for our students.

> single event handler function is a bit involved and fiddly

I think the feeling was that this is a common enough pattern that it would be worth showing.

> little arrow functions are concise

We've gone back and forth on multiple different implementations in this section. I am still unsure what the best option is.

In the past, my hesitation with inline event handlers is that they are somewhat tricky to debug:

```js
<input onClick={event => console.log('foo') setPassword(event.target.value)} />
```

I also think it leads to a bad habit where students dump all the logic inline, leading to poor readability.

I however could see an argument for showing this implementation with separate function definitions first, and mentioning that they could be inlined.

> Maybe this lesson is already long enough

Yes indeed, my perception is that this lesson will already be quite challenging to fit into one day.

My proposal:

1. We switch the first implementation to multiple separate function definitions
2. We then mention that they could be inlined, as with your example
3. Finally, we show a centralised implementation (i.e. the existing one), and mention validation

I think this matches the order of importance of the concepts:

1. Handling multiple form fields
2. Practicing event handling
3. Writing form code concisely
4. Showing patterns for centralised code

Which allows the teacher to cut off if out of time.

Thoughts?

Reviewer: Sounds great! :)
